staging/src/k8s.io/client-go/rest/request.go

```diff
@@ -786,29 +792,45 @@ type WatchListResult struct {
 	// the end of the stream.
 	initialEventsEndBookmarkRV string
 
-	// gv represents the API version
-	// it is used to construct the final list response
-	// normally this information is filled by the server
-	gv schema.GroupVersion
+	// negotiatedObjectDecoder knows how to decode
+	// the initialEventsListBlueprint
+	negotiatedObjectDecoder runtime.Decoder
+
+	// base64EncodedInitialEventsListBlueprint contains an empty,
+	// versioned list encoded in the requested format
+	// (e.g., protobuf, JSON, CBOR) and stored as a base64-encoded string
+	base64EncodedInitialEventsListBlueprint string
 }
 
 func (r WatchListResult) Into(obj runtime.Object) error {
 	if r.err != nil {
 		return r.err
 	}
 
-	listPtr, err := meta.GetItemsPtr(obj)
+	listItemsPtr, err := meta.GetItemsPtr(obj)
 	if err != nil {
 		return err
 	}
-	listVal, err := conversion.EnforcePtr(listPtr)
+	listVal, err := conversion.EnforcePtr(listItemsPtr)
 	if err != nil {
 		return err
 	}
 	if listVal.Kind() != reflect.Slice {
 		return fmt.Errorf("need a pointer to slice, got %v", listVal.Kind())
 	}
 
+	encodedInitialEventsListBlueprint, err := base64.StdEncoding.DecodeString(r.base64EncodedInitialEventsListBlueprint)
+	if err != nil {
+		return err
+	}
+	initialEventsListBlueprint, _, err := r.negotiatedObjectDecoder.Decode(encodedInitialEventsListBlueprint, nil, obj)
+	if err != nil {
+		return err
+	}
+	if initialEventsListBlueprint != obj {
+		return fmt.Errorf("received list type = %T from the server doesn't match the list type = %T", initialEventsListBlueprint, obj)
+	}
```

Comment:
I think this makes sense. Maybe use `runtime.DecodeInto` for consistency with other places with the same expectation?

For regular list requests, if Decode returns an object that is _not_ the same object as "into", with a nil error, then the caller will continue assuming that "into" reflects the object from the REST response, when in fact it has probably not been modified at all: https://github.com/kubernetes/kubernetes/blob/d9c46d8ecb1ede9be30545c9803e17682fcc4b50/staging/src/k8s.io/client-go/rest/request.go#L1535-L1548
IMO, it's dangerous to return a nil error there and it would be worth having a separate discussion about changing that.

Reply:
+1 for using DecodeInto() - it does this check for you for free

Reply:
sorry, i missed the part about `DecodeInto, updated.